sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/meta/provider/pubsub/PubsubIOJsonTable.java

```diff
@@ -0,0 +1,222 @@
+
+/**
+ * <i>Experimental</i>
+ *
+ * <p>Wraps the {@link PubsubIO} with JSON messages into {@link BeamSqlTable}.
+ *
+ * <p>This enables {@link PubsubIO} registration in Beam SQL environment as a table, including DDL
+ * support.
+ *
+ * <p>For example:
+ * <pre>
+ *
+ *  CREATE TABLE topic (name VARCHAR, age INTEGER)
+ *     TYPE 'pubsub'
+ *     LOCATION projects/&lt;GCP project id&gt;/topics/&lt;topic name&gt;
+ *     TBLPROPERTIES '{ \"timestampAttributeKey\" : &lt;timestamp attribute&gt; }';
+ *
+ *   SELECT name, age FROM topic;
+ *
+ * </pre>
+ */
+@AutoValue
+@Internal
+@Experimental
+abstract class PubsubIOJsonTable implements BeamSqlTable, Serializable {
+
+  /**
+   * Schema of the pubsubs message payload.
+   *
+   * <p>Only UTF-8 flat JSON objects are supported at the moment.
+   */
+  abstract Schema getPayloadSchema();
+
+  /**
+   * Attribute key of the Pubsub message from which to extract the event timestamp.
+   *
+   * <p>This attribute has to conform to the same requirements as in {@link
+   * PubsubIO.Read.Builder#withTimestampAttribute}.
+   *
+   * <p>Short version: it has to be either millis since epoch or string in RFC 3339 format.
+   */
+  abstract String getTimestampAttribute();
+
+  /**
+   * Pubsub topic name.
+   *
+   * <p>Topic is the only way to specify the Pubsub source. Explicitly specifying the subscription
+   * is not supported at the moment. Subscriptions are automatically created an managed.
+   */
+  abstract String getTopic();
+
+  static Builder builder() {
+    return new AutoValue_PubsubIOJsonTable.Builder();
+  }
+
+  /**
+   * Table schema.
+   *
+   * <p>Inherited from {@link BeamSqlTable}. Different from {@link #getPayloadSchema()},
+   * includes timestamp attribute.
+   */
+   public abstract Schema getSchema();
+
+  @Override
+  public BeamIOType getSourceType() {
+    return BeamIOType.UNBOUNDED;
+  }
+
+  @Override
+  public PCollection<Row> buildIOReader(Pipeline pipeline) {
+    return
+        PBegin
+            .in(pipeline)
+            .apply("readFromPubsub", readMessagesWithAttributes())
+            .apply("parseTimestampAttribute", parseTimestampAttribute())
+            .apply("parseJsonPayload", parseJsonPayload())
+            .setCoder(getSchema().getRowCoder());
+  }
+
+  private PubsubIO.Read<PubsubMessage> readMessagesWithAttributes() {
+    return
+        PubsubIO
+            .readMessagesWithAttributes()
+            .fromTopic(getTopic())
+            .withTimestampAttribute(getTimestampAttribute());
+  }
+
+  private ParDo.SingleOutput<PubsubMessage, KV<DateTime, PubsubMessage>> parseTimestampAttribute() {
+    return
+        ParDo.of(new DoFn<PubsubMessage, KV<DateTime, PubsubMessage>>() {
+          @ProcessElement
+          public void processElement(ProcessContext context) {
+            PubsubMessage pubsubMessage = context.element();
+            long msSinceEpoch = asMsSinceEpoch(pubsubMessage.getAttribute(getTimestampAttribute()));
```

Comment:
What is the difference between this and 
`msSinceEpoch = context.timestamp().getMillis()` ?

Reply:
You're right. Didn't realize it was already in the context. Will update.